docs/idevicename.1
.TH "idevicename" 1
.SH NAME
idevicename \- Display the device name or set it to NAME if specified.
.SH SYNOPSIS
.B idevicename
[OPTIONS] [NAME]

.SH DESCRIPTION

Simple utility to manage the device name.

If called without any extra argument this tool will print the current device name.

If
.B NAME
is given the device name will be set to the name specified.

.SH OPTIONS
.TP
.B \-u, \-\-udid UDID
target specific device by UDID.
.TP
.B \-n, \-\-network
connect to network device.
.TP
.B \-d, \-\-debug
enable communication debugging.
.TP
.B \-h, \-\-help
prints usage information
.TP
.B \-v, \-\-version
prints version information.

.SH AUTHOR
Nikias Bassen

.SH ON THE WEB
https://libimobiledevice.org

https://github.com/libimobiledevice/libimobiledevice


NEWS
Version 1.3.0
~~~~~~~~~~~~~

* Development release "Let's get the party started."
* Changes:
  - Fix Python 3 support
  - Add more lockdown error codes
  - Add new lockdownd_pair_with_options() function
  - Fix GnuTLS support with iOS 10
  - Make sure sockets only listen locally due to security reasons
  - Plug various memory leaks
  - Fix SSL version negotiation for newer versions of OpenSSL
  - Optimize lockdown pair record handling
  - Return proper error code when a lockdown pair record is missing
  - Fix building with MingGW
  - Store application information in Info.plist using idevicebackup2
  - Fix application backup handling to allow the device to restore applications
    that were installed using idevicebackup2
  - Make idevicebackup2 reboot after restore the default to allow the device to
    migrate data correctly and thus improve the restored device data state
  - Improve console frontend information output in idevicebackup2
  - Extend ideviceprovision tool to allow retrieving and removing all
    provisioning profiles
  - Fix parsing large provisioning profile using ideviceprovision
  - Fix receiving large property lists in property list service
  - Propagate lower level errors to callers instead of returning
    IDEVICE_E_UNKNOWN_ERROR
  - API: Add IDEVICE_DEVICE_PAIRED event type
  - Detect screenshot format to support png, tiff and dat formats using
    idevicescreenshot tool
  - API: Add mobileactivation service implementation
  - Wait for passcode entry if required using idevicesyslog
  - Add HDMI option to diagnostics command for idevicediagnostics
  - Fix IORegistry command for iOS 11+ devices in idevicediagnostics
  - Remove 40-digit character limit for UDID in tools to support newer devices
  - Fix broken validate command in idevicepair with iOS 11+
  - Fix OpenSSL version checks for configure target when using LibreSSL
  - Migrate latest improved common code from libusbmuxd
  - Convert README file to markdown format
  - Fix idevicecrashreport tool to work with iOS 13+
  - Fix various errors in SSL communication logic
  - API: Add preboard service implementation
  - Output hint to user to enter passcode when changing password using
    idevicebackup2
  - Cython: Fix and improve debugserver and diagnostics service bindings
  - API: Add WiFi device support via new idevice_new_with_options() function
  - API: Add idevice_get_device_list_extended() to also list network devices
  - API: Add lockdown_strerror() helper to get error representation as string
  - Add network device support to idevicesyslog and ideviceinfo tools
  - Make debug output consistently output to stderr
  - Add new idevicesetlocation tool (requires mounted developer image)
  - Add option to exit if device disconnects in idevicesyslog
  - API: Add syslog_relay_start_capture_raw() for raw syslog capture
  - Add color output and process filter support to idevicesyslog
  - API: Add companion_proxy service implementation
  - Bump dependency to libusbmuxd 2.0.2
  - Bump dependency to libplist 2.2.0
  - Improve error handling and reporting in library and tools
  - Fix various memory leaks in library and tools
  - Add "--network" and "--version" options to all tools
  - Fix socket_connect_addr() not connecting to network devices using IPv6
    in some cases.
  - Improve IPv6 "scope id" detection to fix connecting to network devices with
    link-local adresses.
  - Update man pages
  - Fix various inconsistent declarations in public headers
  - Allow OpenSSL >= 1.1.0 to use older/disallowed TLS versions fixing issues
    where pairing records were getting removed repeatingly
  - Fixed memory leaks
  - Cython: Rewrite version detection logic in configure.ac
  - Rename "--enable-debug-code" configure option to "--enable-debug"
  - Improve README.md with project description, installation, contributing and
    usage sections
  - Rename library and all related files by adding an API version resulting
    in "libimobiledevice-1.0"
  - Bump soname version
* API is UNSTABLE

Version 1.2.0
~~~~~~~~~~~~~

* Stable release "It took you so long baby!"
* Changes:
  - Require autoconf 2.64 or later
  - Remove dev tools, will return either as proper tools or website examples
  - Refactor installation proxy service implementation and normalize code
  - API: Added instproxy_lookup() to efficiently lookup app information
  - API: Added instproxy_check_capabilities_match() to check device capabilities
  - API: Added various instproxy command and status plist getters
  - API: Make debugserver_client_set_ack_mode() public
  - Fix handling of clients reconnecting in idevicedebugserverproxy which
    previously didn't work properly
  - Flush stdout for every line in idevicesyslog
  - Fix shutdown of idevicedebugserverproxy tool which could hang
  - Notify user when erroneously using idevicebackup with iOS 4 or later
  - Enable build of idevicecrashreport on WIN32
  - Fix thread handle leaks on WIN32 adding thread_new and thread_free
  - cython: Add receive/receive_timeout methods for iDeviceConnection to
    receive raw data from a connection
  - cython: Add new FILE_RELAY_E_PERMISSION_DENIED(-6) error
  - API: Refactor lockdown service internal error checking and add a bunch of
    new native errors
  - Convert int16_t macro error types into enum within common module, too
  - Add new "idevicenotificationproxy" tool to post or observe notifications
  - Fix overlong blocking in np_client_free()
  - Improve maintainability and Requires of pkg-config file
  - API: Add new LOCKDOWN_E_SERVICE_LIMIT error to detect service limit states
  - API: Remove const argv requirement for debugserver_command_new
  - cython: Add get_path_for_bundle_identifier() method to
    InstallationProxyClient
  - cython: Add DebugServerClient class to communicate with debugserver
  - Comply to strict function prototypes by using (void) instead of just ()
  - Fix notification proxy shutdown process which was incorrectly implemented

tools/idevicename.c
/*
 * idevicename.c
 * Simple utility to get or set the device name
 *
 * Copyright (c) 2014  Nikias Bassen, All Rights Reserved.
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 */

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#define TOOL_NAME "idevicename"

#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include <stdlib.h>
#include <getopt.h>
#ifndef WIN32
#include <signal.h>
#endif

#include <libimobiledevice/libimobiledevice.h>
#include <libimobiledevice/lockdown.h>

static void print_usage(void)
{
	printf("Usage: idevicename [OPTIONS] [NAME]\n");
	printf("\n");
	printf("Display the device name or set it to NAME if specified.\n");
	printf("\n");
	printf("OPTIONS:\n");
	printf("  -u, --udid UDID\ttarget specific device by UDID\n");
	printf("  -n, --network\t\tconnect to network device\n");
	printf("  -d, --debug\t\tenable communication debugging\n");
	printf("  -h, --help\t\tprint usage information\n");
	printf("  -v, --version\t\tprint version information\n");
	printf("\n");
	printf("Homepage:    <" PACKAGE_URL ">\n");
	printf("Bug Reports: <" PACKAGE_BUGREPORT ">\n");
}

int main(int argc, char** argv)
{
	int c = 0;
	const struct option longopts[] = {
		{ "udid",    required_argument, NULL, 'u' },
		{ "network", no_argument,       NULL, 'n' },
		{ "debug",   no_argument,       NULL, 'd' },
		{ "help",    no_argument,       NULL, 'h' },
		{ "version", no_argument,       NULL, 'v' },
		{ NULL, 0, NULL, 0}
	};
	int res = -1;
	const char* udid = NULL;
	int use_network = 0;

#ifndef WIN32
	signal(SIGPIPE, SIG_IGN);
#endif

	while ((c = getopt_long(argc, argv, "du:hnv", longopts, NULL)) != -1) {
		switch (c) {
		case 'u':
			if (!*optarg) {
				fprintf(stderr, "ERROR: UDID must not be empty!\n");
				print_usage();
				exit(2);
			}
			udid = optarg;
			break;
		case 'n':
			use_network = 1;
			break;
		case 'h':
			print_usage();
			return 0;
		case 'd':
			idevice_set_debug_level(1);
			break;
		case 'v':
			printf("%s %s\n", TOOL_NAME, PACKAGE_VERSION);
			return 0;
		default:
			print_usage();
			return 2;
		}
	}

	argc -= optind;
	argv += optind;

	if (argc > 1) {
		print_usage();
		return -1;
	}

	idevice_t device = NULL;
	if (idevice_new_with_options(&device, udid, (use_network) ? IDEVICE_LOOKUP_NETWORK : IDEVICE_LOOKUP_USBMUX) != IDEVICE_E_SUCCESS) {
		if (udid) {
			fprintf(stderr, "ERROR: No device found with udid %s.\n", udid);
		} else {
			fprintf(stderr, "ERROR: No device found.\n");
		}
		return -1;
	}

	lockdownd_client_t lockdown = NULL;
	lockdownd_error_t lerr = lockdownd_client_new_with_handshake(device, &lockdown, TOOL_NAME);
	if (lerr != LOCKDOWN_E_SUCCESS) {
		idevice_free(device);
		fprintf(stderr, "ERROR: Could not connect to lockdownd, error code %d\n", lerr);
		return -1;
	}

	if (argc == 0) {
		// getting device name
		char* name = NULL;
		lerr = lockdownd_get_device_name(lockdown, &name);
		if (name) {
			printf("%s\n", name);
			free(name);
			res = 0;
		} else {
			fprintf(stderr, "ERROR: Could not get device name, lockdown error %d\n", lerr);
		}
	} else {
		// setting device name
		lerr = lockdownd_set_value(lockdown, NULL, "DeviceName", plist_new_string(argv[0]));
		if (lerr == LOCKDOWN_E_SUCCESS) {
			printf("device name set to '%s'\n", argv[0]);
			res = 0;
		} else {
			fprintf(stderr, "ERROR: Could not set device name, lockdown error %d\n", lerr);
		}
	}

	lockdownd_client_free(lockdown);
	idevice_free(device);

	return res;
}
